lib/codegen/src/simple_preopt.rs

```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
+    cond_inst_args: ValueList,
+    cond_dest: Ebb,
+    kind: BranchOrderKind,
+}
+
+enum BranchOrderKind {
+    BrzToBrnz,
+    BrnzToBrz,
+    InvertIntCond(IntCC),
+    InvertFloatCond(FloatCC),
+}
+
+fn branch_order(pos: &mut FuncCursor, cfg: &mut ControlFlowGraph, ebb: Ebb, inst: Inst) {
+    let info = match pos.func.dfg[inst] {
+        InstructionData::Jump {
+            opcode: Opcode::Jump,
+            destination,
+            ref args,
+        } => {
+            if let Some(next_ebb) = pos.func.layout.next_ebb(ebb) {
+                if destination == next_ebb {
+                    return;
+                }
+
+                if let Some(prev_inst) = pos.func.layout.prev_inst(inst) {
+                    let prev_inst_data = &pos.func.dfg[prev_inst];
+                    if !prev_inst_data.opcode().is_branch() {
+                        return;
+                    }
+
+                    if let Some(prev_dest) = prev_inst_data.branch_destination() {
+                        if prev_dest != next_ebb {
+                            return;
+                        }
+
+                        match prev_inst_data {
+                            InstructionData::Branch {
+                                opcode,
+                                args: ref prev_args,
+                                destination: cond_dest,
+                                ..
+                            } => {
+                                let cond_arg = {
+                                    let args = pos.func.dfg.inst_args(prev_inst);
+                                    args[0]
+                                };
+
+                                match opcode {
+                                    Opcode::Brz => BranchOrderInfo {
+                                        term_inst: inst,
+                                        term_inst_args: args.clone(),
+                                        term_dest: destination,
+                                        cond_inst: prev_inst,
+                                        cond_arg: cond_arg,
+                                        cond_inst_args: prev_args.clone(),
+                                        cond_dest: *cond_dest,
+                                        kind: BranchOrderKind::BrzToBrnz,
+                                    },
+                                    Opcode::Brnz => BranchOrderInfo {
+                                        term_inst: inst,
+                                        term_inst_args: args.clone(),
+                                        term_dest: destination,
+                                        cond_inst: prev_inst,
+                                        cond_arg: cond_arg,
+                                        cond_inst_args: prev_args.clone(),
+                                        cond_dest: *cond_dest,
+                                        kind: BranchOrderKind::BrnzToBrz,
+                                    },
```

Comment:
Ditto here. The kind is the only thing that's different between the two cases, so you can factor that out.